### PR TITLE
Directory Lister

### PR DESCRIPTION
**Pull request by tdoughty1**
I created a basic directory listing function, this is specifically useful for our experiment, but I hope it is general enough for inclusion.  I followed as closely as I could to the style from list_trees, though I changed it from a dict to a list in order to maintain the file ordering.

Also, sorry for the spam.  Originally, I pushed all the changes except for the generated cpp file, but apparently that breaks continuous integration.  I've now uploaded the appropriate file.

### CODE REVIEW

**Comment by ndawe** [comment]
Thanks a lot for this. Definitely will merge this in. Just a few changes to request.

**Comment by ndawe** [comment]
Also, it would be nice if it listed all nested directories too (along with list_trees). This can be added later in a separate pull request.

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,8 @@ coverage
 .project
 .Rhistory
 .pydevproject
+
+# Unix autogenerated files
+*#
+*~
+
```
Comment (ndawe) [inline .gitignore]: remove this blank line

File: .gitignore [hunk 2]
```diff
@@ -16,3 +16,8 @@ coverage
 .project
 .Rhistory
 .pydevproject
+
```
Comment (ndawe) [inline .gitignore]: remove the two lines here

**Comment by kratsg** [comment]
I would suggest, also, in a future commit, merging `list_directories` and `list_trees` together. They have very, very similar code structure -- the only difference being that you're essentially grouping by class name. One thing that might be possible is to create a hash of lists whose keys are the class names, to make it incredibly easy to retrieve. The first call to a `list_xxxx` function would then cache the results of reading over the structure of the file and successive calls would return a subset of what's been cached?

**Comment by ndawe** [comment]
Yeah, agreed. We are essentially reimplementing rootpy's file.walk() https://github.com/rootpy/rootpy/blob/master/rootpy/io/file.py#L505

**Comment by tdoughty1** [comment]
I've removed the blank lines.  I also took out the comment since it didn't seem to fit with the file.

**Comment by tdoughty1** [comment]
As far as the combination, I'm working on implementing a list_trees function that works with directories.  I still the list directories might be useful on it's own.  That being said I'm not really that familiar with ROOT so I don't really know how most ROOT files are organized.  We use directories, but from what I can tell it seems a little clunky to work with them.

**Comment by tdoughty1** [comment]
I've got a list_trees function that now works with directories, but I was unclear exactly how it should best be output. To start I'm just writing a list with "<dirName>/<treeName>" as I believe this is the format that root2array uses.  Should this be included here?

**Comment by ndawe** [comment]
Let's leave the support for subdirectories for a separate pull request (this will need to be carefully integrated in several places). Paths should only optionally start with /.

**Comment by ndawe** [comment]
I'm merging this. Thanks a lot for the contribution! :+1:

**Pull request merged**